File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/recovery_code/accept/AcceptRecoveryCodePresenter.java
package io.muun.apollo.presentation.ui.recovery_code.accept;


import io.muun.apollo.domain.action.challenge_keys.recovery_code_setup.SetUpRecoveryCodeAction;
import io.muun.apollo.presentation.analytics.AnalyticsEvent;
import io.muun.apollo.presentation.ui.base.SingleFragmentPresenter;
import io.muun.apollo.presentation.ui.base.di.PerFragment;
import io.muun.apollo.presentation.ui.recovery_code.SetupRecoveryCodePresenter;

import android.os.Bundle;
import rx.Observable;

import javax.inject.Inject;

@PerFragment
public class AcceptRecoveryCodePresenter
        extends SingleFragmentPresenter<AcceptRecoveryCodeView, SetupRecoveryCodePresenter> {

    private final SetUpRecoveryCodeAction setUpRecoveryCode;

    @Inject
    public AcceptRecoveryCodePresenter(SetUpRecoveryCodeAction setUpRecoveryCode) {
        this.setUpRecoveryCode = setUpRecoveryCode;
    }

    @Override
    public void setUp(Bundle arguments) {
        super.setUp(arguments);

        view.setTexts(userSel.get());

        setUpUpdateChallengeSetupAction();
    }

    private void setUpUpdateChallengeSetupAction() {
        final Observable<?> observable = setUpRecoveryCode
                .getState()
                .compose(handleStates(view::setLoading, this::handleError))
                .doOnNext(ignored -> getParentPresenter().onSetupSuccessful());

        subscribeTo(observable);
    }

    /**
     * Send Houston our new challenge key.
     */
    public void finishSetup() {
        setUpRecoveryCode.run(getParentPresenter().getRecoveryCode().toString());
    }

    public void showAbortDialog() {
        getParentPresenter().showAbortDialog();
    }

    @Override
    protected AnalyticsEvent getEntryEvent() {
        return new AnalyticsEvent.S_FINISH_RECOVERY_CODE_CONFIRM();
    }
}

File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/recovery_code/accept/AcceptRecoveryCodePresenter.kt
package io.muun.apollo.presentation.ui.recovery_code.accept

import android.os.Bundle
import io.muun.apollo.domain.action.challenge_keys.recovery_code_setup.FinishRecoveryCodeSetupAction
import io.muun.apollo.domain.analytics.AnalyticsEvent
import io.muun.apollo.domain.analytics.AnalyticsEvent.S_FINISH_RECOVERY_CODE_CONFIRM
import io.muun.apollo.presentation.ui.base.SingleFragmentPresenter
import io.muun.apollo.presentation.ui.base.di.PerFragment
import io.muun.apollo.presentation.ui.recovery_code.SetupRecoveryCodePresenter
import javax.inject.Inject

@PerFragment
internal class AcceptRecoveryCodePresenter @Inject constructor(
    private val finishRecoveryCodeSetup: FinishRecoveryCodeSetupAction,
) : SingleFragmentPresenter<AcceptRecoveryCodeView, SetupRecoveryCodePresenter>() {

    override fun setUp(arguments: Bundle) {
        super.setUp(arguments)
        view.setTexts(userSel.get())

        finishRecoveryCodeSetup
            .state
            .compose(handleStates(view::setLoading, this::handleError))
            .doOnNext {
                parentPresenter.onSetupSuccessful()
            }
            .let(this::subscribeTo)
    }

    /**
     * Finish/Verify in Houston our new challenge key setup.
     */
    fun finishSetup() {
        finishRecoveryCodeSetup.run()
    }

    fun showAbortDialog() {
        parentPresenter.showAbortDialog()
    }

    override fun getEntryEvent(): AnalyticsEvent {
        return S_FINISH_RECOVERY_CODE_CONFIRM()
    }
}